syncopy/shared/metadata.py

```diff
@@ -0,0 +1,262 @@
+    """
+    if details is None:
+        return
+
+    close_file = False
+    if isinstance(h5fout, str):
+        close_file = True # We openend it, we close it.
+        h5fout = h5py.File(h5fout, mode="w")
+
+    if isinstance(unique_key_suffix, int):
+        unique_key_suffix = "__" + str(unique_key_suffix) + "_0"
+
+    grp = h5fout['metadata'] if 'metadata' in h5fout else h5fout.create_group("metadata")
+    attribs = _parse_backend_metadata(details)
+    for k, v in attribs.items():
+        k_unique = k + unique_key_suffix
+        grp.attrs.create(k_unique, data=v)
+    h5fout.flush()
+
+    if close_file:
+        h5fout.close()
+
+
+def metadata_trial_indices_abs(metadata, selection):
+    """
+    Re-compute absolute trial indices in the metadata from the relative ones.
+
+    Note that the input metadata is already preprocessed from the hdf5.
+    """
+    if metadata is None:
+        return metadata  # Nothing to do.
+    if selection is None or selection.trials is None:
+        return metadata  # abs_trial_idx = rel_trial_idx, so nothing to do for us.
+    del_keys = list()
+
+    metadata_to_add = dict()
+    for unique_md_label_rel, v in metadata.items():
+        label, rel_trial_idx, call_idx = decode_unique_md_label(unique_md_label_rel)
+        rel_trial_idx = int(rel_trial_idx)
+        abs_trial_idx = selection.trials[rel_trial_idx]
+        if abs_trial_idx != rel_trial_idx:
+            unique_md_label_abs = encode_unique_md_label(label, abs_trial_idx, call_idx)
+            metadata_to_add[unique_md_label_abs] = v  # Re-add value with new key later (cannot add in iteration).
+            del_keys.append(unique_md_label_rel)
+
+    for k in del_keys:
+        del metadata[k]
+    metadata = { **metadata,  **metadata_to_add}
+    return metadata
+
+
+def encode_unique_md_label(label, trial_idx, call_idx=0):
+    """Assemble something like `test`, `2` and `0` into `test__2_0`."""
+    return(label + "__" + str(trial_idx) + "_" + str(call_idx))
+
+
+def decode_unique_md_label(unique_label):
+    """
+    Splits something like `test__2_0` into `test`, `2` and `0`.
+
+    Parameters
+    ----------
+    unique_label: str, with format `<label>__<trial_idx>_<chunk_idx>'`.
+
+    Returns
+    -------
+    tuple of `str`: the `'label'`, `'trial_idx'` and `'chunk_idx'`.
+    """
+    lab_ind = unique_label.rsplit("__")
+    label = lab_ind[0]
+    trialidx_callidx = lab_ind[1].rsplit("_")
+    return label, trialidx_callidx[0], trialidx_callidx[1]
+
+
+def extract_md_group(md):
+    """
+    Extract metadata from h5py 'metadata' group and return a standard dict.
+
+    Parameters
+    ----------
+    md: a h5py group, that contains metadata attributes as 'attrs'.
+
+    Returns
+    -------
+    dict, containing entries of type `(str, np.ndarray)`.
+    """
+    metadata = dict()
+    for k, v in md.attrs.items():
+        metadata[k] = v.copy() # copy the numpy array
```

Comment:
the `h5file['metadata']` is a hdf5 group, not a dict. even though it has a dict-like interface. it will vanish if you close the file.

I gotta say I am not 100% sure whether the `.copy()` is really needed though.

Reply:
Ok, but as these attr arrays are anyways <=64kB, I guess we don't loose any substantial RAM here :+1: